multi.c

```diff
@@ -0,0 +1,145 @@
+#include "multi.h"
+
+#include "brute.h"
+#include "queue.h"
+#include "single.h"
+#include <assert.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+void *
+mt_password_check (void *context)
+{
+  mt_context_t *mt_context = (mt_context_t *)context;
+  task_t task;
+  st_context_t st_context = {
+    .hash = mt_context->config->hash,
+    .data = { .initialized = 0 },
+  };
+
+  while (true)
+    {
+      /* don't understand why we need status check here */
+      if (queue_pop (&mt_context->queue, &task) == S_FAILURE)
+        {
+          return NULL;
+        }
+      if (st_password_check (&task, &st_context))
+        {
+          /* This is printing but loop doesn't break */
+          /* printf ("Password found: %s\n", task.password); */
+          memcpy (mt_context->password, task.password, sizeof (task.password));
+        }
+
+      if (pthread_mutex_lock (&mt_context->mutex) != 0)
+        {
+          return NULL;
+        }
+      --mt_context->passwords_remaining;
+      /* printf("-- %d\n", mt_context->passwords_remaining); */
+      if (mt_context->passwords_remaining == 0)
+        {
+          if (pthread_cond_signal (&mt_context->cond_sem) != 0)
+            {
+              return NULL;
+            }
+        }
+      if (pthread_mutex_unlock (&mt_context->mutex) != 0)
+        {
+          return NULL;
+        }
+    }
+  return NULL;
+}
+
+bool
+queue_push_wrapper (task_t *task, void *context)
+{
+  mt_context_t *mt_context = (mt_context_t *)context;
+  if (queue_push (&mt_context->queue, task) == S_FAILURE)
+    {
+      return false;
+    }
+
+  if (pthread_mutex_lock (&mt_context->mutex) != 0)
```

Comment:
Here is a really hard race condition. Code will work in 99.99% cases, but will stall at exit once in a while. 
This is what I call "violation of transaction semantics".
1. First you need to declare in log journal planned transaction.
2. Next you need to make a transaction.
3. Finally you need to remove from journal completed transaction.

In this particular example it means that you need to increase the password_remaining first, then push password to queue, then pop password from queue, check password and finally decrease password_remaining. Increase/decrease of password_remaining is a logging in journal, push/pop/check - transaction. I.e. sequence should be increase-push-pop-check-decrease.


With the current order you may get sequence: push-pop-check-decreas-increase and password_remaining will not hit 0, so signal to main thread will not be emitted.